src/main/kotlin/no/nav/syfo/client/azuread/AzureAdClient.kt

```diff
@@ -1,80 +1,90 @@
+        val cachedToken = cache.get(key = cacheKey)
+        return if (cachedToken?.isExpired() == false) {
+            cachedToken
+        } else {
+            val azureAdTokenResponse = getAccessToken(
+                Parameters.build {
+                    append("client_id", clientId)
+                    append("client_secret", clientSecret)
+                    append("grant_type", "client_credentials")
+                    append("scope", "api://$scopeClientId/.default")
+                }
+            )
+            azureAdTokenResponse?.let { token ->
+                token.toAzureAdToken().also {
+                    cache[cacheKey] = it
+                }
+            }
+        }
+    }
 
-            val response = httpClientWithProxy.post(aadAccessTokenUrl) {
+    private suspend fun getAccessToken(
+        formParameters: Parameters,
+    ): AzureAdTokenResponse? =
+        try {
+            val response: HttpResponse = httpClient.post(aadAccessTokenUrl) {
                 accept(ContentType.Application.Json)
-
-                setBody(
-                    FormDataContent(
-                        Parameters.build {
-                            append("client_id", clientId)
-                            append("scope", resource)
-                            append("grant_type", "client_credentials")
-                            append("client_secret", clientSecret)
-                        },
-                    ),
-                )
-            }
-            when (response.status) {
-                HttpStatusCode.OK ->
-                    tokenMap[resource] = response.body()
-                HttpStatusCode.BadGateway ->
-                    throw BadRequestException(exceptionErrorMessage("Bad request - $resource"))
-                HttpStatusCode.InternalServerError ->
-                    throw InternalServerErrorException(exceptionErrorMessage("Internal server error - $response"))
-                else ->
-                    throw RuntimeException(exceptionErrorMessage("Exception - $response"))
+                setBody(FormDataContent(formParameters))
             }
+            response.body<AzureAdTokenResponse>()
+        } catch (e: ResponseException) {
+            handleUnexpectedResponseException(e)
+            null
         }
-        return tokenMap[resource]!!.access_token
+
+    private fun handleUnexpectedResponseException(
+        responseException: ResponseException,
+    ) {
+        log.error(
+            "Error while requesting AzureAdAccessToken with statusCode=${responseException.response.status.value}",
+            responseException
+        )
     }
 
-    private fun exceptionErrorMessage(msg: String) = "Could not get token from AzureAD: $msg"
+    companion object {
+        const val CACHE_AZUREAD_TOKEN_SYSTEM_KEY_PREFIX = "azuread-token-system-"
+        private val cache = ConcurrentHashMap<String, AzureAdToken>()
```

Comment:
Hvordan funker denne cachen egentlig? Vil ikke en hashmap uansett leve i scope av den som kaller det, og ikke på tvers av kall? Burde vi hatt noe redis her?

Reply:
(og også mtp flere pods?)

Reply:
Dette er kopiert fra isaktivitetskrav. Det vil vel bli opprettet én instans av AzureAdClient per pod, og derfor en cache per pod. Vet ikke om vi trenger cache på tvers av pod'ene?

Reply:
Men den er singleton så cache treffer på tvers, in-memory?

Ikke sikkert vi trenger på tvers av pod, men kan vel være fare for at vi må cache flere ganger da :) Ikke sikkert det er noe stort problem uansett, de bruker client-side cache i modia frontend også

Reply:
Det opprettes én instans av AzureAdClient i App.kt, så da mener jeg at det skal bli bare én cache per pod.